### PR TITLE
feat: Ipfs implementation

### PR DESCRIPTION
**Pull request by tommasini**
**Description**
More info [here](https://github.com/MetaMask/mobile-planning/issues/1162)
Videos are attached to the link above

**Screenshots/Recordings**

_If applicable, add screenshots and/or recordings to visualize the before and after of your change_

**Issue**

fixes #???

**Checklist**

* [ ] There is a related GitHub issue
* [ ] Tests are included if applicable
* [ ] Any added code is fully documented

### CODE REVIEW

File: app/components/Views/BrowserTab/index.js [hunk 1]
```diff
@@ -490,7 +490,21 @@ export const BrowserTab = (props) => {
           Logger.error(err, 'Failed to resolve ENS name');
         }
 
-        Alert.alert(strings('browser.failed_to_resolve_ens_name'), err.message);
+        if (
+          err?.message?.startsWith(
+            'IPFS gateway is disabled on security and privacy settings',
```
Comment (sethkfman) [inline app/components/Views/BrowserTab/index.js:495]: Can you make this a constant and reference it where the error originates and here? This way if the message changes the code won't break. If not, I would add a test for this condition and validate the behavior so we can know if it needs to be changed.

File: app/components/Views/Settings/SecuritySettings/Sections/DeleteMetaMetricsData.tsx [hunk 2]
```diff
@@ -125,6 +125,7 @@ const DeleteMetaMetricsData = () => {
 
   return (
     <SettingsButtonSection
+      testID="delete-metrics-button"
```
Comment (sethkfman) [inline app/components/Views/Settings/SecuritySettings/Sections/DeleteMetaMetricsData.tsx:128]: I think these test ID need should be constants to they can be referenced in E2E

File: app/component-library/components/Avatars/Avatar/variants/AvatarToken/AvatarToken.tsx [hunk 3]
```diff
@@ -28,6 +31,7 @@ const AvatarToken = ({
     isHaloEnabled,
     showFallback,
   });
+  const isIpfsGatewayEnabled = useSelector(selectIsIpfsGatewayEnabled);
```
[the remark this answers is not in the document]
Comment (tommasini) [reply app/component-library/components/Avatars/Avatar/variants/AvatarToken/AvatarToken.tsx:34]: Are you referring to this logic? This function could and should be on an utils file, but not sure if It was this you was mentioning 
```
  const isIpfsDisabledAndUriIsIpfs = imageSource
    ? !isIpfsGatewayEnabled && isIPFSUri(imageSource)
    : false;

```

File: app/components/Views/BrowserTab/index.js (hunk 1 above)
Comment (tommasini) [reply app/components/Views/BrowserTab/index.js]: THanks Seth! Already addressed !

File: app/lib/ens-ipfs/resolver.js [hunk 4]
```diff
@@ -51,7 +57,11 @@ export default async function resolveEnsToIpfsContentId({ provider, name }) {
     const contentId = multihash.toB58String(
       multihash.encode(buffer, 'sha2-256'),
     );
-
+    if (!Engine.context.PreferencesController.state.isIpfsGatewayEnabled) {
+      throw new Error(
+        `IPFS gateway is disabled on security and privacy settings`,
```
[the remark this answers is not in the document]
Comment (tommasini) [reply app/lib/ens-ipfs/resolver.js]: Addressed!

File: app/components/Views/Settings/SecuritySettings/Sections/DeleteMetaMetricsData.tsx (hunk 2 above)
Comment (tommasini) [reply app/components/Views/Settings/SecuritySettings/Sections/DeleteMetaMetricsData.tsx:128]: Thanks Seth! Already addressed!

File: app/component-library/components/Avatars/Avatar/variants/AvatarToken/AvatarToken.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (sethkfman) [reply app/component-library/components/Avatars/Avatar/variants/AvatarToken/AvatarToken.tsx:34]: After discussion this component and IPFS toggle it is correctly being used

**Review by sethkfman — DISMISSED** [review]
LGTM

**Review by sethkfman — APPROVED** [review]
LGTM

**Review by cortisiko — APPROVED** [review]
Confirmed the dev test cases are good. ✅

**Pull request merged**